### PR TITLE
Adds Encryptable model behavior

### PR DESCRIPTION
**Pull request by der-On**
This pull requests adds the Encryptable model behavior which is based on the Encryptable model trait but can be added dynamically using the models static `::extend` method.

Usage:

```php
MyModelClass::extend(function ($model) {
  $model->addDynamicMethod('getEncryptableAttributes', function () {
    return ['private_attribute'];
  });
  $model->implement[] = \Winter\Storm\Database\Behaviors\Encryptable::class;
});
```

### CODE REVIEW

**Comment by mjauvin** [comment]
@der-On can you look into the failed tests?

**Comment by bennothommo** [comment]
@der-On one thing I've noticed is that during the `bootEncryptable()` method, you're doing a check to see if the `$encryptable` property exists in the model, and throwing an exception if it doesn't exist. The problem is - your comment above stipulates that the `$encryptable` property should be protected (like it is with the Encryptable trait), but unfortunately this won't work in the context of a behavior, because they cannot access protected or private properties and methods.

Also, your example has the `getEncryptableAttributes()` method dynamically added, which would make the `$encryptable` property redundant.

So to make this work, either that property check needs to be dropped, or the `$encryptable` property needs to be made public. Ideally, it would be better DX if they don't have to create a dynamic method every time they wish to use a behavior, but on the flip-side, I'm not sure allowing the `$encryptable` property to be public is a good idea.

If my [scoped and local extensions PR](https://github.com/wintercms/storm/pull/134) is merged, we'd have a mechanism to be able to allow us to access protected properties, so this could be a good push to get that merged.

I'll commit a test and my changes which demonstrate the behavior working when `$encryptable` is public, but ultimately we need to make a decision on how we want to go forward with that scenario first.

**Comment by der-On** [comment]
@LukeTowers @bennothommo wow thanks for all the input and improvements. This is becoming interesting.

**Comment by LukeTowers** [comment]
@bennothommo what's left on this?

**Comment by bennothommo** [comment]
Given that the extension changes were merged, just converting these changes in this PR to use that instead.

**Comment by LukeTowers** [comment]
@bennothommo is that something that you are able to take care of or are you wanting @der-On to do that or is @mjauvin interested in doing that (he's currently assigned to the PR?).

**Comment by bennothommo** [comment]
If someone else can tackle it sooner, feel free, otherwise I'll circle to it sometime in the near future.

File: src/Database/Behaviors/Encryptable.php [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        $this->model = $parent;
+        $this->bootEncryptable();
+    }
+
+    /**
+     * Boot the encryptable trait for a model.
+     */
+    public function bootEncryptable(): void
+    {
+        if (!$this->model->propertyExists('encryptable')) {
```
Comment (mjauvin) [inline src/Database/Behaviors/Encryptable.php]:
```suggestion
        $isEncryptable = $this->model->extend(function () {
            return $this->propertyExists('encryptable');
        });
        if (!$isEncryptable) {
```

File: src/Database/Behaviors/Encryptable.php [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+                'You must define an $encryptable property on the %s class to use the Encryptable behavior.',
+                get_class($this->model)
+            ));
+        }
+
+        /*
+         * Encrypt required fields when necessary
+         */
+        $this->model->bindEvent('model.beforeSetAttribute', function ($key, $value) {
+            if (in_array($key, $this->model->getEncryptableAttributes()) && !is_null($value)) {
```
Comment (mjauvin) [inline src/Database/Behaviors/Encryptable.php]:
```suggestion
            if (in_array($key, $this->getEncryptableAttributes()) && !is_null($value)) {
```

File: src/Database/Behaviors/Encryptable.php [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+                return $this->model->makeEncryptableValue($key, $value);
```
Comment (mjauvin) [inline src/Database/Behaviors/Encryptable.php]:
```suggestion
                return $this->makeEncryptableValue($key, $value);
```

File: src/Database/Behaviors/Encryptable.php [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+            }
+        });
+        $this->model->bindEvent('model.beforeGetAttribute', function ($key) {
+            if (in_array($key, $this->model->getEncryptableAttributes()) && array_get($this->model->getAttributes(), $key) != null) {
```
Comment (mjauvin) [inline src/Database/Behaviors/Encryptable.php]:
```suggestion
            if (in_array($key, $this->getEncryptableAttributes()) && array_get($this->model->getAttributes(), $key) != null) {
```

File: src/Database/Behaviors/Encryptable.php [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    /**
+     * Decrypts an attribute value
+     */
+    public function getEncryptableValue(string $key): mixed
+    {
+        $attributes = $this->model->getAttributes();
+        return isset($attributes[$key])
+            ? $this->model->getEncrypter()->decrypt($attributes[$key])
```
Comment (mjauvin) [inline src/Database/Behaviors/Encryptable.php]:
```suggestion
            ? $this->getEncrypter()->decrypt($attributes[$key])
```

File: src/Database/Behaviors/Encryptable.php [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+            : null;
+    }
+
+    /**
+     * Returns a collection of fields that will be encrypted.
+     */
+    public function getEncryptableAttributes(): array
+    {
+        return $this->model->encryptable;
```
Comment (mjauvin) [inline src/Database/Behaviors/Encryptable.php]:
```suggestion
        return $this->model->extend(function () {
            return $this->encryptable;
        });
```

File: tests/Database/Behaviors/EncryptableBehaviorTest.php [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    }
+}
+
+class TestModelEncryptableBehavior extends \Winter\Storm\Database\Model
+{
+    public $implement = [
+        \Winter\Storm\Database\Behaviors\Encryptable::class,
+    ];
+
+    public $encryptable = ['secret'];
```
Comment (mjauvin) [inline tests/Database/Behaviors/EncryptableBehaviorTest.php]:
```suggestion
    protected $encryptable = ['secret'];
```

File: src/Database/Behaviors/Encryptable.php [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+         * Encrypt required fields when necessary
+         */
+        $this->model->bindEvent('model.beforeSetAttribute', function ($key, $value) {
+            if (in_array($key, $this->model->getEncryptableAttributes()) && !is_null($value)) {
+                return $this->model->makeEncryptableValue($key, $value);
+            }
+        });
+        $this->model->bindEvent('model.beforeGetAttribute', function ($key) {
+            if (in_array($key, $this->model->getEncryptableAttributes()) && array_get($this->model->getAttributes(), $key) != null) {
+                return $this->model->getEncryptableValue($key);
```
Comment (mjauvin) [inline src/Database/Behaviors/Encryptable.php]:
```suggestion
                return $this->getEncryptableValue($key);
```

File: src/Database/Behaviors/Encryptable.php [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+        });
+    }
+
+    /**
+     * Encrypts an attribute value and saves it in the original locker.
+     */
+    public function makeEncryptableValue(string $key, mixed $value): string
+    {
+        $this->originalEncryptableValues[$key] = $value;
+        return $this->model->getEncrypter()->encrypt($value);
```
Comment (mjauvin) [inline src/Database/Behaviors/Encryptable.php]:
```suggestion
        return $this->getEncrypter()->encrypt($value);
```

**Review by mjauvin — CHANGES_REQUESTED** [review]

File: src/Database/Behaviors/Encryptable.php (hunk 1 above)
Comment (der-On) [reply src/Database/Behaviors/Encryptable.php]: @mjauvin Will `$this` actually be available in the closure?

File: src/Database/Behaviors/Encryptable.php (hunk 9 above)
Comment (der-On) [reply src/Database/Behaviors/Encryptable.php]: Done

File: src/Database/Behaviors/Encryptable.php (hunk 6 above)
Comment (der-On) [reply src/Database/Behaviors/Encryptable.php]: @mjauvin Will `$this` actually be available in the closure?

File: tests/Database/Behaviors/EncryptableBehaviorTest.php (hunk 7 above)
Comment (der-On) [reply tests/Database/Behaviors/EncryptableBehaviorTest.php]: @mjauvin If the above closures you suggested will work, this can be done.

Comment (mjauvin) [reply tests/Database/Behaviors/EncryptableBehaviorTest.php]: It should work with latest develop branch for sure.

**Comment by der-On** [comment]
@mjauvin Made all the changes. Let's see what the tests say.

**Comment by der-On** [comment]
@mjauvin Seems like there is some unknown error with the code analysis but unit tests have passed.

**Comment by mjauvin** [comment]
@bennothommo how do we solve this with phpstan, we're accessing a protected property with the help of the `extend()` method...

**Comment by bennothommo** [comment]
Surprised that the tests are passing - I think this branch is still missing all the changes to the Extension framework.

**Comment by bennothommo** [comment]
Should be all good now.

**Comment by der-On** [comment]
Awesome! Sorry for me taking so long to get back to this.

File: src/Database/Behaviors/Encryptable.php [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+     */
+    public function bootEncryptable(): void
+    {
+        $isEncryptable = $this->model->extend(function () {
+            /** @var Model $this */
+            return $this->propertyExists('encryptable');
+        });
+
+        if (!$isEncryptable) {
+            throw new Exception(sprintf(
```
Comment (mjauvin) [inline src/Database/Behaviors/Encryptable.php]:
```suggestion
            throw new ApplicationException(sprintf(
```

File: src/Database/Behaviors/Encryptable.php [hunk 11]
```diff
@@ -0,0 +1,158 @@
+<?php namespace Winter\Storm\Database\Behaviors;
+
+use App;
+use Exception;
```
Comment (mjauvin) [inline src/Database/Behaviors/Encryptable.php]:
```suggestion
use ApplicationException;
```

File: src/Database/Behaviors/Encryptable.php [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+    {
+        return $this->originalEncryptableValues;
+    }
+
+    /**
+     * Returns the original values of any encrypted attributes.
+     */
+    public function getOriginalEncryptableValue(string $attribute): mixed
+    {
+        return $this->originalEncryptableValues[$attribute] ?? null;
```
Comment (mjauvin) [inline src/Database/Behaviors/Encryptable.php]:
```suggestion
        return array_get($this->originalEncryptableValues, $attribute, null);
```

File: src/Database/Behaviors/Encryptable.php [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+        /*
+         * Encrypt required fields when necessary
+         */
+        $this->model->bindEvent('model.beforeSetAttribute', function ($key, $value) {
+            if (in_array($key, $this->getEncryptableAttributes()) && !is_null($value)) {
+                return $this->makeEncryptableValue($key, $value);
+            }
+        });
+        $this->model->bindEvent('model.beforeGetAttribute', function ($key) {
+            if (in_array($key, $this->getEncryptableAttributes()) && array_get($this->model->getAttributes(), $key) != null) {
```
Comment (mjauvin) [inline src/Database/Behaviors/Encryptable.php]:
```suggestion
            if (in_array($key, $this->getEncryptableAttributes()) && array_get($this->model->attributes, $key) != null) {
```

**Review by mjauvin — CHANGES_REQUESTED** [review]

**Review by mjauvin — APPROVED** [review]

File: src/Database/Behaviors/Encryptable.php [hunk 14]
```diff
@@ -1,7 +1,7 @@
 <?php namespace Winter\Storm\Database\Behaviors;
 
 use App;
-use Exception;
+use ApplicationException;
```
Comment (mjauvin) [inline src/Database/Behaviors/Encryptable.php]:
```suggestion
use Winter\Storm\Exception\ApplicationException;
```

**Comment by mjauvin** [comment]
Looks good to me!

**Review by bennothommo — APPROVED** [review]

**Comment by der-On** [comment]
Whohoo! Thanks.

**Pull request merged**